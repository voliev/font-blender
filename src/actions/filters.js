import {
  SET_CATEGORIES_FILTER,
  SET_SORTING_ORDER,
  SET_SORTING_PARAM,
  SET_TITLE_FILTER
} from "../constants";

export const setTitleFilter = (title = "") => ({
  type: SET_TITLE_FILTER,
  title
});

export const setCategoriesFilter = category => ({
  type: SET_CATEGORIES_FILTER,
  category
});

export const setSortingParam = param => ({
  type: SET_SORTING_PARAM,
  param
});

export const setSortingOrder = order => ({
  type: SET_SORTING_ORDER,
  order
});
