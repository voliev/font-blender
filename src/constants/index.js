// *** ACTION TYPES ***

// Background
export const SET_BACKGROUND_COLOR = "SET_BACKGROUND_COLOR";

// Filteres
export const SET_CATEGORIES_FILTER = "SET_CATEGORIES_FILTER";
export const SET_SORTING_ORDER = "SET_SORTING_ORDER";
export const SET_SORTING_PARAM = "SET_SORTING_PARAM";
export const SET_TITLE_FILTER = "SET_TITLE_FILTER";

// Fonts
export const FETCH_FONTS_DATA = "FETCH_FONTS_DATA";
export const FETCH_FONTS_DATA_SUCCESS = "FETCH_FONTS_DATA_SUCCESS";
export const FETCH_FONTS_DATA_FAILURE = "FETCH_FONTS_DATA_FAILURE";
export const SET_FONTS = "SET_FONTS";
export const UPDATE_FONT = "UPDATE_FONT";

// Notifications
export const ADD_NOTIFICATION = "ADD_NOTIFICATION";
export const REMOVE_NOTIFICATION = "REMOVE_NOTIFICATION";

// Showcase
export const DISPLAY_SHOWCASE_TEXT = "DISPLAY_SHOWCASE_TEXT";
export const DISPLAY_STYLES = "DISPLAY_STYLES";
export const DISPLAY_SHOWCASE_FONT_PREVIEW = "DISPLAY_SHOWCASE_FONT_PREVIEW";

// Styles
export const SET_DEFAULT_STATE = "SET_DEFAULT_STATE";
export const UNDO_RESET = "UNDO_RESET";

// Text styles
export const SET_STYLES_INHERITANCE = "SET_STYLES_INHERITANCE";
export const UPDATE_ELEMENT_STYLES = "UPDATE_ELEMENT_STYLES";

// Selected fonts
export const ADD_SELECTED_FONT = "ADD_SELECTED_FONT";
export const CLEAR_SELECTED_FONTS = "CLEAR_SELECTED_FONTS";
export const REMOVE_FONT_FROM_SELECTED = "REMOVE_FONT_FROM_SELECTED";
export const UNDO_DESELECTED = "UNDO_DESELECTED";

// Type scale
export const SET_DEFAULT_TYPE_SCALE_STATE = "SET_DEFAULT_TYPE_SCALE_STATE";
export const SET_FONT_SIZE = "SET_FONT_SIZE";
export const SET_TYPE_SCALE = "SET_TYPE_SCALE";

// *** MISCELLANEOUS ***

// Google Fonts API endpoint URL
export const GOOGLE_FONTS_API_ENDPOINT =
  "https://www.googleapis.com/webfonts/v1/webfonts";
// Fonts' bundle size that should be prefetched
export const BUNDLE_SIZE = 50;
// Available text elements for styling
export const ELEMENTS = ["body", "h1", "h2", "h3", "h4", "li", "small"];
// Font's character range that needs to be prefetched
export const CHAR_RANGE =
  "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
// Font showcase string
export const SHOWCASE_STRING =
  "That's the beauty of music. They can't get that from you...";
// Default showcase font size
export const DEFAULT_FONT_SIZE = 16;
// Available type scale values
export const SCALE_VALUES = {
  1.067: "Minor Second",
  1.125: "Major Second",
  1.2: "Minor Third",
  1.25: "Major Third",
  1.333: "Perfect Fourth",
  1.414: "Augmented Fourth",
  1.5: "Perfect Fifth",
  1.618: "Golden Ratio"
};
// Minimum gap between the color picker icon and viewport
// edge required to place the color picker
export const MIN_GAP = 300;
// Notification snackbar visibility timeout
export const TIMEOUT = 10000;
