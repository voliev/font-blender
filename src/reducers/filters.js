import {
  SET_CATEGORIES_FILTER,
  SET_TITLE_FILTER,
  SET_SORTING_ORDER,
  SET_SORTING_PARAM
} from "../constants";

const initialState = {
  title: "",
  categories: {
    serif: true,
    "sans-serif": true,
    display: true,
    monospace: true,
    handwriting: true
  },
  sorting: {
    param: "popularity",
    order: "asc"
  }
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_TITLE_FILTER:
      return {
        ...state,
        title: action.title
      };
    case SET_CATEGORIES_FILTER:
      return {
        ...state,
        categories: {
          ...state.categories,
          [action.category]: !state.categories[action.category]
        }
      };
    case SET_SORTING_PARAM:
      return {
        ...state,
        sorting: {
          ...state.sorting,
          param: action.param
        }
      };
    case SET_SORTING_ORDER:
      return {
        ...state,
        sorting: {
          ...state.sorting,
          order: action.order === "asc" ? "des" : "asc"
        }
      };
    default:
      return state;
  }
};

export const getFilters = state => state.filters;
