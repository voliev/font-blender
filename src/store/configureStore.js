import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import fonts from "../reducers/fonts";
import filters from "../reducers/filters";
import selectedFonts from "../reducers/selectedFonts";
import showcase from "../reducers/showcase";
import styles from "../reducers/styles";
import notifications from "../reducers/notifications";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default () => {
  const store = createStore(
    combineReducers({
      fonts,
      filters,
      selectedFonts,
      showcase,
      styles,
      notifications
    }),
    composeEnhancers(applyMiddleware(thunk))
  );

  return store;
};
