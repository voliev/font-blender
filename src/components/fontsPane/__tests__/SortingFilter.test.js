import React from "react";
import { mount, shallow } from "enzyme";
import configureStore from "redux-mock-store";
import SortingFilterConnected, { SortingFilter } from "../SortingFilter";
import { SET_SORTING_ORDER, SET_SORTING_PARAM } from "../../../constants";
import state from "../../../fixtures/state";
import filters from "../../../fixtures/filters";

const setup = (renderFn = shallow, propsUpdates = {}, connected = false) => {
  if (connected) {
    const mockStore = configureStore();
    const store = mockStore(state);

    return {
      store,
      wrapper: renderFn(<SortingFilterConnected store={store} />)
    };
  }

  const props = {
    sorting: filters.sorting,
    setSortingParam: jest.fn(),
    setSortingOrder: jest.fn(),
    ...propsUpdates
  };

  return {
    ...props,
    wrapper: renderFn(<SortingFilter {...props} />)
  };
};

afterEach(() => {
  jest.clearAllMocks();
});

it("should render without crashing", () => {
  const { wrapper } = setup();

  expect(wrapper).toMatchSnapshot();
});

it("should receive corresponding props", () => {
  const { sorting, setSortingOrder, setSortingParam, wrapper } = setup(mount);

  expect(typeof wrapper.prop("sorting")).toBe("object");
  expect(wrapper.prop("sorting")).toEqual(sorting);

  expect(typeof wrapper.prop("setSortingParam")).toBe("function");
  expect(wrapper.prop("setSortingParam")).toBe(setSortingParam);

  expect(typeof wrapper.prop("setSortingOrder")).toBe("function");
  expect(wrapper.prop("setSortingOrder")).toBe(setSortingOrder);
});

it("should handle sorting parameter change", () => {
  const { wrapper, setSortingParam } = setup();

  wrapper.instance().handleSortingParamChange({
    target: {
      value: "alpha"
    }
  });

  expect(setSortingParam).toHaveBeenCalledTimes(1);
  expect(setSortingParam).toHaveBeenCalledWith("alpha");
});

it("should handle sorting parameter change", () => {
  const { wrapper, setSortingOrder } = setup();

  wrapper.instance().handleSortingOrderChange({
    currentTarget: {
      value: "des"
    }
  });

  expect(setSortingOrder).toHaveBeenCalledTimes(1);
  expect(setSortingOrder).toHaveBeenCalledWith("des");
});

it("should dispatch an action", () => {
  const { wrapper, store } = setup(mount, undefined, true);

  wrapper.find(SortingFilter).prop("setSortingParam")("alpha");

  const actions = store.getActions();

  expect(actions).toContainEqual({
    type: SET_SORTING_PARAM,
    param: "alpha"
  });
});

it("should dispatch an action", () => {
  const { wrapper, store } = setup(mount, undefined, true);

  wrapper.find(SortingFilter).prop("setSortingOrder")("des");

  const actions = store.getActions();

  expect(actions).toContainEqual({
    type: SET_SORTING_ORDER,
    order: "des"
  });
});
