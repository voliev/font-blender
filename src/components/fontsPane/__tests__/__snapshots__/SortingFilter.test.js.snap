// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<div
  className="SortingFilter"
>
  <SortingParameters
    sortingParamChangeHandler={[Function]}
    value="popularity"
  />
  <SortingOrder
    sortingOrderChangeHandler={[Function]}
    value="asc"
  />
</div>
`;
